feat: Add gradient style option

TheGameProfi opened the pull request:
This pull request enhances the `renderCard` function in `src/utils/renderCard.tsx` by introducing support for gradient backgrounds and improving background handling. The changes include updating logic for background assignment, adding support for all other background options alongside gradient options, and modifying the card's styling to use the new background property.

### Enhancements to background handling:
* Introduced a `background` variable in `renderCard` to handle both solid color and gradient backgrounds.
* Modified the card's inline styles to use the `background` property instead of `backgroundColor`.
* Updating bg option to support either hex code with and without `#` and usage of any other background css option

See following Issue => #69

TheGameProfi committed:
fix(card): add Gradient Style Option

cnrad commented on src/components/ProfileCard.tsx:
```diff
@@ -139,7 +143,7 @@ export const ProfileCard: React.FC<ProfileCardProps> = ({
             height: `${divHeight}px`,
             inset: 0,
-            backgroundColor: `#${backgroundColor}`,
+            background: background,
```
shorthand

cnrad reviewed (changes requested):
Sorry for the delay - here's what I think would work best, and it's a little tricky since we need to avoid using hashes:

Let's keep the `background` parameter and not worry about adding another parameter for `gradient` - rather, it can use the background value as given. That is, you could have `?background=2500dc`  *or* `?background=linear-gradient(45deg,f00,0f0)` and it will work both ways. If the length of the `background` parameter is 3 or 6 (for pure hex codes), append a `#` to the beginning if it is. If not, split by comma and then run the same check (length is 3 or 6) and append the `#` if necessary. With this method, they should be able to use `rgb` or `rgba` as well, just without the commas.

After all that, if you don't mind updating the readme to be in line with this too, that'd be great. Make sure to specify that hex codes should be used without the `#`

cnrad commented:
> The gradient syntax allows for a lot of variation, and reliably parsing and modifying it to prepend # could easily lead to edge cases or unexpected behavior.

Could you list some examples of edge cases? My current idea is this: 
- parse for `linear-gradient(` and `)`
- whatever is enclosed, split by comma
- check the length of each segment and if it's 3 or 6, append the `#`
- join back together

As of now, I don't see how that could really go wrong (assuming users will follow the README for syntax - if they don't, skill issue). Either way, your original code doesn't handle hashes, so the styling will break for anything after the gradient in the query params anyways. I wish things were as easy as they seem

> The color of the border around the online status also uses the background parameter

This is a good point that I forgot about completely - however, there are ways to create a mask around elements to let the background naturally act as a border, see [this demo](https://play.tailwindcss.com/PsQQVdkd5t). This is probably a bit complex though, so it's understandable if you don't want to tackle that

Let me know your thoughts on all of this

TheGameProfi commented:
Yeah, I currently know of two cases off the top of my head that would need a bit of special handling:

* `linear-gradient(145deg, 103204 10%, 10%, blue)`

  * The `145deg` part would incorrectly trigger the hex check, but that’s fixable with an extra condition to filter out angles.
  * The `103204 10%` part wouldn’t trigger the check at all since it's longer than a valid hex code for which we would need a special checking to maybe split again for spaces?

> *"Either way, your original code doesn't handle hashes, so the styling will break for anything after the gradient in the query params anyways."*

Could you clarify what you mean by that? Not sure I fully follow. Are you referring to how the query string is parsed or something else?

> *"There are ways to create a mask around elements to let the background naturally act as a border."*

That sounds interesting. I’ll look into that approach. If I run into issues trying to get it working, I’ll reach out.

TheGameProfi commented:
Ah, that’s what you meant. I didn’t know it was called that.
There’s actually a solution already used in the landing UI where you can configure the Card. It uses URL encoding: https://www.w3schools.com/tags/ref\_urlencode.ASP
That replaces special chars with valid URL text. For example, `bg=#400` becomes `bg=%23400`. We also need it anyway for whitespaces in the gradient option.

nilscrafthd commented:
Nice

TheGameProfi committed:
refactor(card): remove aditional gradient option
- add function to check if hex code
- add handling for hex with/without #
- support for all background option that are url encoded

TheGameProfi commented:
No comments since 7 months.
Bumping @cnrad
